pallets/parachain-staking/src/set.rs

```diff
@@ -16,89 +16,117 @@
 };
-
 #[cfg(feature = "std")]
-use serde::{Deserialize, Serialize};
+use sp_std::{fmt, prelude::*};
 
-/// An ordered set backed by `Vec`.
-#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
-#[derive(RuntimeDebug, PartialEq, Eq, Encode, Decode, Default, Clone)]
-pub struct OrderedSet<T>(Vec<T>);
+/// An ordered set backed by `BoundedVec`.
+#[derive(PartialEq, Eq, Encode, Decode, DefaultNoBound, Clone)]
+pub struct OrderedSet<T, S>(pub BoundedVec<T, S>);
 
-impl<T: Ord> OrderedSet<T> {
+impl<T: Ord + Clone, S: Get<u32>> OrderedSet<T, S> {
 	/// Create a new empty set.
 	pub fn new() -> Self {
-		Self(Vec::new())
+		Self(BoundedVec::default())
 	}
 
-	/// Create an ordered set from a `Vec`.
+	/// Creates an ordered set from a `BoundedVec`.
 	///
 	/// The vector will be sorted reversily (from greatest to lowest) and
 	/// deduped first.
-	pub fn from(mut v: Vec<T>) -> Self {
+	pub fn from(bv: BoundedVec<T, S>) -> Self {
+		let mut v = bv.into_inner();
 		v.sort_by(|a, b| b.cmp(a));
 		v.dedup();
-		Self::from_sorted_set(v)
+		Self::from_sorted_set(v.try_into().expect("No values were added"))
 	}
 
-	/// Create a set from a `Vec`.
+	/// Create a set from a `BoundedVec`.
 	///
 	/// Assumes that `v` is sorted reversely (from greatest to lowest) and only
 	/// contains unique elements.
-	pub fn from_sorted_set(v: Vec<T>) -> Self {
-		Self(v)
+	pub fn from_sorted_set(bv: BoundedVec<T, S>) -> Self {
+		Self(bv)
 	}
 
-	/// Insert an element, if no equal item exist in the set.
+	/// Inserts an element, if no equal item exist in the set.
+	///
+	/// Throws if insertion would exceed the bounded vec's max size.
 	///
-	/// Return true if the item is unique in the set, otherwise returns false.
-	pub fn insert(&mut self, value: T) -> bool {
+	/// Returns true if the item is unique in the set, otherwise returns false.
+	pub fn insert(&mut self, value: T) -> Result<bool, ()> {
 		match self.linear_search(&value) {
-			Ok(_) => false,
+			Ok(_) => Ok(false),
 			Err(loc) => {
-				self.0.insert(loc, value);
-				true
+				self.0.try_insert(loc, value)?;
+				Ok(true)
 			}
 		}
 	}
 
-	/// Insert or replaces an element.
+	/// Attempts to replace the last element of the set with the provided value.
+	/// Assumes the set to have reached its bounded size.
 	///
-	/// Returns the old value if existing.
-	pub fn upsert(&mut self, value: T) -> Option<T> {
+	/// Throws with `false` if the value already exists in the set.
+	/// Throws with `true` if the value has the least order in the set, i.e.,
+	/// it would be appended (inserted at i == length).
+	///
+	/// Returns the replaced element upon success.
+	pub fn try_insert_replace(&mut self, value: T) -> Result<T, bool> {
+		let last_idx = self.len().saturating_sub(1);
+		match self.linear_search(&value) {
+			Ok(_) => Err(false),
+			Err(i) if i < self.len() => {
+				// always replace the last element
+				let old = sp_std::mem::replace(&mut self.0[last_idx], value);
+				self.sort_greatest_to_lowest();
```

Comment:
couldn't we just pop the last entry and insert at the specific index?
We could but we would need to handle the error case for the insert then. That's unfortunate.